src/main/java/org/springframework/retry/support/RetryTemplate.java

```diff
@@ -296,6 +296,10 @@ protected <T, E extends Throwable> T doExecute(RetryCallback<T, E> retryCallback
 				throw new TerminatedRetryException("Retry terminated abnormally by interceptor before first attempt");
 			}
 
+			if (!context.hasAttribute(RetryContext.MAX_ATTEMPTS)) {
```

Comment:
potentially we could only set the value in the context when different from NO_MAXIMUM_ATTEMPTS_SET so that in the context it will be either  null or maxAttempts rather than NO_MAXIMUM_ATTEMPTS_SET or maxAttempts - probably the latter approach could be cleaner